src/storage/src/hummock/compactor/mod.rs

```diff
@@ -559,57 +577,68 @@ impl Compactor {
         if options.estimate_bloom_filter_capacity == 0 {
             options.estimate_bloom_filter_capacity = options.capacity / DEFAULT_ENTRY_SIZE;
         }
-        let builder_factory = RemoteBuilderFactory {
-            sstable_id_manager: self.context.sstable_id_manager.clone(),
-            limiter: self.context.read_memory_limiter.clone(),
-            options,
-            remote_rpc_cost: get_id_time.clone(),
-            filter_key_extractor,
-        };
 
-        // NOTICE: should be user_key overlap, NOT full_key overlap!
-        let mut builder = CapacitySplitTableBuilder::new(
-            builder_factory,
-            self.cache_policy,
-            self.sstable_store.clone(),
-            self.context.stats.clone(),
-        );
+        macro_rules! make_builder_and_compact {
+            ($writer_builder:expr) => {{
+                // Monitor time cost building shared buffer to SSTs.
+                let compact_timer = if self.context.is_share_buffer_compact {
+                    self.context.stats.write_build_l0_sst_duration.start_timer()
+                } else {
+                    self.context.stats.compact_sst_duration.start_timer()
+                };
+
+                let mut builder = self.get_multi_builder(
+                    $writer_builder,
+                    options,
+                    get_id_time.clone(),
+                    filter_key_extractor,
+                );
+
+                Compactor::compact_and_build_sst(
+                    &mut builder,
+                    &self.key_range,
+                    iter,
+                    self.gc_delete_keys,
+                    self.watermark,
+                    compaction_filter,
+                )
+                .await?;
+                let ret = builder.finish()?;
+                compact_timer.observe_duration();
+                ret
+            }};
+        }
 
-        // Monitor time cost building shared buffer to SSTs.
-        let compact_timer = if self.context.is_share_buffer_compact {
-            self.context.stats.write_build_l0_sst_duration.start_timer()
+        let split_table_outputs = if self.context.options.enable_sst_streaming_upload {
+            let writer_builder = get_writer_builder_for_streaming_upload(
+                self.sstable_store.clone(),
+                self.cache_policy,
+            );
+            make_builder_and_compact!(writer_builder)
```

Comment:
We can even pass `get_writer_builder_for_streaming_upload` and `get_writer_builder_for_batch_upload` as the input to `make_builder_and_compact` to avoid more code duplication.